chore: update packages setup eslint and simplify prettier config as we use default

edouard-lopez opened the pull request:
* puppeteer to 9x → 21.x
* typescript 4.x → 5.x
* ts-node 9.x → 10.x
* jest 26.x → 29.x
* jest-image-snapshot 4.x → 6.x

edouard-lopez committed:
chore: update packages setup eslint and simplify prettier config as we use default

* puppeteer to 9x → 21.x
* typescript 4.x → 5.x
* ts-node 9.x → 10.x
* jest 26.x → 29.x
* jest-image-snapshot 4.x → 6.x

OmarTawfik commented on src/index.ts:
```diff
@@ -12,6 +12,7 @@ export async function renderScreenshot(options: Partial<TerminalScreenshotOption
 
   const templatePath = await generateTemplate(validatedOtions);
   const browser = await puppeteer.launch({
+    headless: "new",
```
is that needed/left over from #10?

```suggestion
```

edouard-lopez replied:
Nope, it's a way to mute this Puppeteer warning:
>       Puppeteer old Headless deprecation warning:                                                                                                                                                    
>         In the near future `headless: true` will default to the new Headless mode                                                                                                                    
>         for Chrome instead of the old Headless implementation. For more                                                                                                                              
>         information, please see https://developer.chrome.com/articles/new-headless/.                                                                                                                 
>         Consider opting in early by passing `headless: "new"` to `puppeteer.launch()`                                                                                                                
>         If you encounter any bugs, please report them to https://github.com/puppeteer/puppeteer/issues/new/choose.

OmarTawfik commented on src/cli.ts:
```diff
@@ -28,15 +28,18 @@ const options = new Command("terminal-screenshot")
+    if (error instanceof Error) {
+      // eslint-disable-next-line no-console
+      console.error("Error: " + error.message);
+    }
```
If it was not a native `Error` instance, let's still log it?

```suggestion
    } else {
          console.error(error);
    }
```

OmarTawfik replied:
Let's also log the error in case it wasn't a native `Error` instance:

```suggestion
    } else {
      console.error(error);
    }
```

OmarTawfik commented on .vscode/settings.json:
```diff
@@ -1,10 +1,13 @@
   "editor.defaultFormatter": "esbenp.prettier-vscode",
   "editor.formatOnSave": true,
-  "editor.rulers": [120],
+  "editor.rulers": [
```
nit: is this formatting intended?
this file (and `tsconfig.json`) now fail `prettier` validation.

edouard-lopez replied:
I got not difference

```
❯ prettier --version
3.2.4
```

```
❯ prettier --list-different  ./tsconfig.json .vscode/settings.json 
tsconfig.json
.vscode/settings.json

```
What version are you using?

OmarTawfik replied:
looks like this is another hickup with setup/CI. I fixed in #18
Apologies for that :) the code hasn't been updated in a few years. Rebasing should make this go away.

OmarTawfik commented on .eslintrc:
```diff
@@ -0,0 +1,25 @@
+{
```
if we are adding `eslint`, then I suggest adding it to `yarn lint` command, to enforce it for CI/other users as well: `yarn eslint . --max-warnings 0`

edouard-lopez replied:
done

OmarTawfik commented on .eslintrc:
```diff
@@ -0,0 +1,25 @@
+        "plugin:@typescript-eslint/recommended"
+    ],
+    "rules": {
+        "no-console": 2, // Remember, this means error!
```
Explicit `error` works for me. I wonder are you using the integer here?

```suggestion
        "no-console": "error",
```

edouard-lopez replied:
Thanks for the feedback, got updated on various things :+1: 

I copy/pasted (:scream:)  from another project to get something working quickly

OmarTawfik commented on .eslintrc:
```diff
@@ -0,0 +1,25 @@
+    "rules": {
+        "no-console": 2, // Remember, this means error!
+        "camelcase": "warn",
+        "@typescript-eslint/no-explicit-any": "warn",
```
https://typescript-eslint.io/rules/no-explicit-any/

> Extending "plugin:@typescript-eslint/recommended" in an ESLint configuration enables this rule.

I wonder why are we adding it here?

OmarTawfik commented on .eslintrc:
```diff
@@ -0,0 +1,25 @@
+    ],
+    "rules": {
+        "no-console": 2, // Remember, this means error!
+        "camelcase": "warn",
```
https://typescript-eslint.io/rules/camelcase/

> This rule has been deprecated in favour of the naming-convention rule.

OmarTawfik commented on .eslintrc:
```diff
@@ -0,0 +1,25 @@
+        "no-console": 2, // Remember, this means error!
+        "camelcase": "warn",
+        "@typescript-eslint/no-explicit-any": "warn",
+        "@typescript-eslint/no-unused-vars": [
```
https://typescript-eslint.io/rules/no-unused-vars/

> Extending "plugin:@typescript-eslint/recommended" in an ESLint configuration enables this rule.


Also, `^_` is the default for `argsIgnorePattern`. Am I reading this right?

OmarTawfik commented on package.json:
```diff
@@ -33,25 +33,28 @@
+    "@typescript-eslint/eslint-plugin": "^6.20.0",
+    "@typescript-eslint/parser": "^6.20.0",
     "chalk": "4.1.1",
+    "eslint": "^8.56.0",
```
I suggest using explicit versions for all of these, instead of the `^` operator, matching the rest of them.

OmarTawfik replied:
Looks like one is still using it.

OmarTawfik commented on package.json:
```diff
@@ -33,25 +33,28 @@
-    "@types/jest-image-snapshot": "4.3.0",
-    "@types/puppeteer": "5.4.3",
+    "@types/jest": "29.5.11",
+    "@types/jest-image-snapshot": "6.4.0",
```
I suggest keeping these lists alphabetically sorted.

OmarTawfik replied:
I think it is still unsorted. If you would like, you can use the VS Code `Sort lines ascending` command after selecting the lines in question.

OmarTawfik commented on package.json:
```diff
@@ -33,25 +33,28 @@
     "joi": "17.4.0",
-    "puppeteer": "9.1.1",
-    "xterm": "4.12.0"
+    "xterm": "4.12.0",
```
The only non-internal/tooling change here is for `puppeteer`. I wonder if we should do this in #10 instead? since it is a product change, is only used with/impacts `xterm`. Otherwise, I'm fine with keeping the existing version if it works well.

edouard-lopez replied:
will keep here unless necessary

OmarTawfik replied:
Sounds good. Thanks.

OmarTawfik reviewed (changes requested):
Thank you. Left a few concerns/suggestions.

edouard-lopez committed:
chore(eslint): update based on review

edouard-lopez committed:
fix: log the error in case it wasn't a native Error instance

edouard-lopez committed:
chore(package): fix version and sort

edouard-lopez committed:
fix: correct build error TS2403: AbortSignal' was also declared here.

$ tsc
Error: node_modules/@types/node/globals.d.ts(341,13): error TS2403: Subsequent variable declarations must have the same type.  Variable 'AbortSignal' must be of type '{ new (): AbortSignal; prototype: AbortSignal; abort(reason?: any): AbortSignal; timeout(milliseconds: number): AbortSignal; }', but here has type '{ new (): AbortSignal; prototype: AbortSignal; }'.

edouard-lopez commented:
I fixed the formatting issue, my IDE was using the JSON formatter instead of prettier on `tsconfig.json` and `setttings.json`.

OmarTawfik commented on package.json:
```diff
@@ -11,7 +11,7 @@
   },
   "scripts": {
-    "lint": "prettier . --check",
+    "lint": "eslint --max-warnings 0 ./ && prettier --check ./",
```
Looks like this PR also added 3 outdated/legacy images to `tests/screenshots/`, perhaps from a rebase. I suggest removing them.

OmarTawfik commented on yarn.lock:
```diff
@@ -2,302 +2,357 @@
 # yarn lockfile v1
 
 
```
This PR changes the entire `yarn.lock` from `yarnpkg` registry to `npmjs` .. I wonder if you have custom local overrides (e.g. `yarnrc`). If not, I suggest reverting these changes, and undoing changes to this file and rerunning `yarn` with vanilla settings.

edouard-lopez committed:
chore: sort package

edouard-lopez committed:
chore: fix yarn.lock registry

OmarTawfik commented:
Thank you for the effort!
